README.md
# NAME

Nephia::Plugin::Teng - Simple ORMapper Plugin For Nephia

# SYNOPSIS

    use Nephia plugins => [qw/Teng/];

    path '/person/:id' => sub {
        my $id = path_param('id');
        my $row = teng->lookup('person', { id => $id });
        return res { 404 } unless $row;

        return {
            id => $id,
            name => $row->get_column('name'),
            age => $row->get_column('age'),
        };
    };

Read row from person table in database in this code.

# DESCRIPTION

## configuration - configuration for Teng.

configuration file:

    'Plugin::Teng' => {
        connect_info => ['dbi:SQLite:dbname=data.db'],
        plugins => [qw/Lookup Pager/]
    },

The "connect\_info" is connect information for [DBI](http://search.cpan.org/perldoc?DBI).

Enumerate in "plugins" option if you want load Teng plugins.

## teng - Create Teng Object

"teng" DSL create the Teng Object.

## database\_do - load SQL before plackup.

In this example to create table before plackup.

in controller :

    database_do "CREATE TABLE IF NOT EXISTS person (id INTEGER, name TEXT, age INTEGER);"

    path '/' => sub {
        ...
    };

# SEE ALSO

[Nephia](http://search.cpan.org/perldoc?Nephia)

[Teng](http://search.cpan.org/perldoc?Teng)

# LICENSE

Copyright (C) macopy.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

mackee <macopy123\[attttt\]gmai.com>

ichigotake


Changes
Revision history for Perl extension Nephia-Plugin-Teng

{{$NEXT}}

0.04 2013-09-30T14:33:48Z
    - !!NOT COMPATIBLE VOSON JUMP!! support to Nephia-0.80 or latter

0.03 2013-07-13T01:43:53Z

    - support for Nephia-0.31 higher

0.02 2013-07-10T11:17:01Z

    - fix example code in POD

0.01 2013-07-08T15:06:14Z

    - original version



t/nephia-test_app/etc/conf/common.pl
### common config
+{
    'Plugin::Teng' => {
        connect_info => [ 'DBI:SQLite:dbname=data.db' ],
        plugins => [qw/Lookup/],
    }
};


t/nephia-test_app/lib/Nephia/TestApp.pm
package Nephia::TestApp;
use strict;
use warnings;
use Nephia plugins => [qw/Teng/];
use utf8;

database_do <<'SQL';
CREATE TABLE IF NOT EXISTS `person` (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    name TEXT,
    age INTEGER
);
SQL

database_do <<'SQL';
INSERT INTO "person"
    ("id", "name", "age")
    VALUES (1, "bob", "20");
SQL


get '/person/:id' => sub {
    my $id = path_param('id');
    return res { 422 } unless $id =~ /^[0-9]+$/;

    my $row = teng->lookup('person', { id => $id });
    return res { 404 } unless $row;

    return {
        id => $id,
        name => $row->get_column('name'),
        age => $row->get_column('age'),
    };
};

post '/register' => sub {
    my $name = param->{name};
    return res { 422 } unless $name;
    my $age = param->{age};
    return res { 422 } unless $age =~ /^[0-9]+$/;

    my $id = teng->fast_insert('person', { name => $name, age => $age });

    return {
        id => $id,
        name => $name,
        age => $age,
    };
};

1;

=head1 NAME

MyApp - Web Application

=head1 SYNOPSIS

  $ plackup

=head1 DESCRIPTION

MyApp is web application based Nephia.

=head1 AUTHOR

clever guy

=head1 SEE ALSO

Nephia

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Nephia-Plugin-Teng',
    module_name     => 'Nephia::Plugin::Teng',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Nephia/Plugin/Teng.pm
package Nephia::Plugin::Teng;
use 5.008005;
use strict;
use warnings;
use Teng::Schema::Loader;
use DBI;
use parent 'Nephia::Plugin';

our $VERSION = "0.03";

sub new {
    my ($class, %opts) = @_;
    my $self = $class->SUPER::new(%opts);
    $self->{RUN_SQL} = [];
    return $self;
}

sub exports { qw/database_do teng/ }

sub database_do {
    my $self = shift;
    my $context = shift;
    $self->_load_context_config($context);
    return sub ($) {
        my $sql = shift;
        push @{$self->{RUN_SQL}}, $sql;
    };
}

sub _on_connect_do {
    my $self = shift;
    my $dbh = $self->_create_dbh();
    my @RUN_SQL = @{$self->{RUN_SQL}};
    for my $sql (@RUN_SQL) {
        $dbh->do($sql);
    }
}

sub teng {
    my $self = shift;
    my $context = shift;
    $self->_load_context_config($context);
    return sub (@) {
        $self->{teng} ||= $self->_create_teng();
    };
}

sub _create_teng {
    my $self = shift;
    my $config = $self->{opts};
    my $pkg = $config->{teng_namespace} // $self->app->{caller}.'::DB';

    $self->_on_connect_do();

    my $teng =
        Teng::Schema::Loader->load(
            dbh => $self->_create_dbh(),
            namespace => $pkg
        );

    for my $plugin (@{$config->{teng_plugins}}) {
        $pkg->load_plugin($plugin);
    }

    return $teng;
};

sub _create_dbh {
    my $self = shift;
    my @connect_info;

    @connect_info = @{$self->{opts}->{connect_info}}
        if exists $self->{opts}->{connect_info}
            && ref $self->{opts}->{connect_info} eq "ARRAY";
    @connect_info = @{$self->{config}->{DBI}->{connect_info}}
        if !@connect_info
            && exists $self->{config}->{DBI}->{connect_info}
            && ref $self->{config}->{DBI}->{connect_info} eq "ARRAY";

    return DBI->connect(@connect_info);
}

sub _load_context_config {
    my $self = shift;
    my $context = shift;
    $self->{config} = $context->get('config');
}

1;
__END__

=encoding utf-8

=head1 NAME

Nephia::Plugin::Teng - Simple ORMapper Plugin For Nephia

=head1 SYNOPSIS

    use Nephia plugins => [qw/Teng/];

    path '/person/:id' => sub {
        my $id = path_param('id');
        my $row = teng->lookup('person', { id => $id });
        return res { 404 } unless $row;

        return {
            id => $id,
            name => $row->get_column('name'),
            age => $row->get_column('age'),
        };
    };

Read row from person table in database in this code.

=head1 DESCRIPTION

=head2 configuration - configuration for Teng.

configuration file:

    'DBI' => {
        connect_info => ['dbi:SQLite:dbname=data.db'],
        teng_plugins => [qw/Lookup Pager/]
    },

The "connect_info" is connect information for L<DBI>.

Enumerate in "plugins" option if you want load Teng plugins.

=head2 teng - Create Teng Object

"teng" DSL create the Teng Object.

=head2 database_do - load SQL before plackup.

In this example to create table before plackup.

in controller :

    database_do "CREATE TABLE IF NOT EXISTS person (id INTEGER, name TEXT, age INTEGER);"

    path '/' => sub {
        ...
    };

=head1 SEE ALSO

L<Nephia>

L<Teng>

=head1 LICENSE

Copyright (C) macopy.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

mackee E<lt>macopy123[attttt]gmai.comE<gt>

ichigotake

=cut

